### PR TITLE
Access to django views. Backend Middleware. Generate Permissions based on ACL allowed list

### PR DESCRIPTION
**Pull request by ntodosiychuk**
Updated CustomViewProcessMiddleware class with new functional(uses function for generate view permission name. Check user permission id database). 
Updated README.rst.

### CODE REVIEW

File: view_acl/constants.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from django.conf import settings
+# List with constant variables.
+PERM_PREFIX = 'view_'
+DESC_PREFIX = 'Custom '
+
+URLCONF = getattr(settings, 'ROOT_URLCONF', ())
+ACL_APPS = getattr(settings, 'ACL_APPS', ())
```
Comment (pzdeb) [inline view_acl/constants.py]: Let it be ACL_ALLOWED_VIEWS

File: view_acl/constants.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+from django.conf import settings
+# List with constant variables.
+PERM_PREFIX = 'view_'
+DESC_PREFIX = 'Custom '
+
+URLCONF = getattr(settings, 'ROOT_URLCONF', ())
```
Comment (pzdeb) [inline view_acl/constants.py]: Do we need this variable?

File: view_acl/utils.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    return permissions
+
+def generate_perm_name(func):
+    """ Generate permission name from callback function. """
+    return '.'.join((func and func.__module__ or '',
+                     func and func.__name__ or ''))
+
+def refresh_permissions(permissions):
+    """ Refresh permission:
+        Remove deprecated permissions,
```
Comment (pzdeb) [inline view_acl/utils.py]: Please fix formatting of this docstring

File: view_acl/constants.py (hunk 1 above)
Comment (ntodosiychuk) [inline view_acl/constants.py]: Ok

File: view_acl/constants.py (hunk 2 above)
Comment (ntodosiychuk) [inline view_acl/constants.py]: I think we can remove it... We use it only in one place

File: view_acl/utils.py (hunk 3 above)
Comment (ntodosiychuk) [inline view_acl/utils.py]: Ok

File: view_acl/__init__.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    """ Autodiscover for urls.py """
+    permissions_to_add = {}
+    # Get permissions based on urlpatterns from urls.py
+    resolver = urlresolvers.get_resolver(URLCONF)
+    urlpatterns = resolver.url_patterns
+    permissions = generate_permissions(urlpatterns)
+
+    # Prepare data for refreshing permissions.
+    for perm in permissions:
+        permissions_to_add[PERM_PREFIX + perm] = DESC_PREFIX + perm
```
Comment (pzdeb) [inline view_acl/__init__.py]: I couldn't find where you use values of `permissions_to_add` with the `DESC_PREFIX`.

And why this prefix starts with "Custom " by default?

Comment (ntodosiychuk) [inline view_acl/__init__.py]: (DESC_PREFIX + perm_name) its field name in Permission model.
'Custom' we add to distinguish that this perm is generated manualy.

**Comment by ntodosiychuk** [comment]
@pzdeb @pivancy 
Coverage:
constants           100%
tests/test_utils    100%
utils                     78%   48-63

File: setup.py [hunk 5]
```diff
@@ -25,4 +26,8 @@
         'Topic :: Internet :: WWW/HTTP',
         'Topic :: Internet :: WWW/HTTP :: Dynamic Content',
     ],
-)
\ No newline at end of file
+)
+
+if sys.argv[-1] == 'test':
```
Comment (pzdeb) [inline setup.py]: This section ideally must be part of setuptools. Or at least moved into `if __name__ == '__main__'`

File: view_acl/__init__.py (hunk 4 above)
Comment (pzdeb) [inline view_acl/__init__.py]: Manually? But you really add that permissions automatically from autodiscover, no?

File: view_acl/tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+        utils.ACL_ALLOWED_VIEWS = ('views_test.home',)
+
+        urlpatterns = [
+            url(r'^$', views_test.home),
+            url(r'^test/$', views_test.home),
+            url(r'^home/$', views_test.home)]
+
+        perms = generate_permissions(urlpatterns, [])
+        self.assertListEqual(['views_test.home',
+                              'views_test.home',
```
Comment (pzdeb) [inline view_acl/tests/test_utils.py]: As result we can have only one unique permission in the list of generated permissions

File: view_acl/tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+from utils import generate_permissions, generate_perm_name, is_acl_apps
+
+
+class TestUtils(unittest.TestCase):
+
+    def test_generate_permissions(self):
+        utils.ACL_ALLOWED_VIEWS = ('views_test*',)
+        urlpatterns = [url(r'^$', views_test.index, name='index')]
+        perms = generate_permissions(urlpatterns)
+        self.assertListEqual(['views_test.index'], perms)
```
Comment (pzdeb) [inline view_acl/tests/test_utils.py]: In constants module you have PERM_PREFIX as `view_`, so how can this test match plural form?

File: view_acl/tests/test_utils.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+from django.contrib import admin
+
+from utils import generate_permissions, generate_perm_name, is_acl_apps
+
+
+class TestUtils(unittest.TestCase):
+
+    def test_generate_permissions(self):
+        utils.ACL_ALLOWED_VIEWS = ('views_test*',)
+        urlpatterns = [url(r'^$', views_test.index, name='index')]
```
Comment (pzdeb) [inline view_acl/tests/test_utils.py]: Don't you use `name` parameter for permission codename generation?

File: view_acl/tests/views_test.py [hunk 9]
```diff
@@ -0,0 +1,22 @@
+from django.http import HttpResponse
+
+
+def home(request):
+    return HttpResponse('home')
```
Comment (pzdeb) [inline view_acl/tests/views_test.py]: What is the idea for this test?

File: view_acl/utils.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+def generate_perm_name(func):
+    """Generate permission name from callback function."""
+    return '.'.join((func and func.__module__ or '',
+                     func and func.__name__ or ''))
+
+
+def is_acl_apps(perm):
+    """Check if permission is on acl list."""
+    for app in ACL_ALLOWED_VIEWS:
```
Comment (pzdeb) [inline view_acl/utils.py]: `for view in ...`

File: view_acl/utils.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+            if is_acl_apps(perm):
+                permissions.append(perm)
+        elif isinstance(pattern, urlresolvers.RegexURLResolver):
+            generate_permissions(pattern.url_patterns, permissions)
+    return permissions
+
+
+def generate_perm_name(func):
+    """Generate permission name from callback function."""
+    return '.'.join((func and func.__module__ or '',
```
Comment (pzdeb) [inline view_acl/utils.py]: I can't find where you handle application name. Does that newly generated permission contain application name?

File: view_acl/utils.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        if isinstance(pattern, urlresolvers.RegexURLPattern):
+            perm = generate_perm_name(pattern.callback)
+            if is_acl_apps(perm):
+                permissions.append(perm)
+        elif isinstance(pattern, urlresolvers.RegexURLResolver):
+            generate_permissions(pattern.url_patterns, permissions)
+    return permissions
+
+
+def generate_perm_name(func):
```
Comment (pzdeb) [inline view_acl/utils.py]: > generate_codename()

File: view_acl/utils.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    """Generate permission name from callback function."""
+    return '.'.join((func and func.__module__ or '',
+                     func and func.__name__ or ''))
+
+
+def is_acl_apps(perm):
```
Comment (pzdeb) [inline view_acl/utils.py]: please rename the function to `allowed_permission()` or `is_allowed_permission()`

File: view_acl/tests/test_utils.py [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+        utils.ACL_ALLOWED_VIEWS = ('views',)
+        assert is_acl_apps('views_test.results') is None
+
+        utils.ACL_ALLOWED_VIEWS = ('',)
+        assert is_acl_apps('views_test.results') is None
+
+        utils.ACL_ALLOWED_VIEWS = ('',)
+        assert is_acl_apps('views_test.results') is None
+
+        utils.ACL_ALLOWED_VIEWS = ('*',)
```
Comment (pzdeb) [inline view_acl/tests/test_utils.py]: Anyway we have to use some predefined exclusion list for wildcard cases. We can't block native Django application such as `admin, auth` and others

File: view_acl/tests/test_utils.py (hunk 6 above)
Comment (ntodosiychuk) [inline view_acl/tests/test_utils.py]: Ok, probably better to use set here. 
But I'm sure that we don't create duplicates in database.

Comment (pzdeb) [inline view_acl/tests/test_utils.py]: definitely we don't and also can't generate duplicates in the code

File: view_acl/tests/test_utils.py (hunk 7 above)
Comment (ntodosiychuk) [inline view_acl/tests/test_utils.py]: But we add this prefix after generating permission name.
In this case 'views_test' is module, just name

File: view_acl/tests/test_utils.py (hunk 8 above)
Comment (ntodosiychuk) [inline view_acl/tests/test_utils.py]: Unfortunately no, we couldn't handle case when we have two or more the same name and view_name. In this case the  generated name is not Unique

File: view_acl/tests/views_test.py (hunk 9 above)
Comment (ntodosiychuk) [inline view_acl/tests/views_test.py]: I need only callback func for test, for check correct name generation.

File: view_acl/utils.py (hunk 10 above)
Comment (ntodosiychuk) [inline view_acl/utils.py]: will be fixed ;)

File: view_acl/utils.py (hunk 11 above)
Comment (ntodosiychuk) [inline view_acl/utils.py]: Yes, we app is on func.__module__ attribute

File: view_acl/utils.py (hunk 12 above)
Comment (ntodosiychuk) [inline view_acl/utils.py]: Agree!

File: view_acl/tests/test_utils.py (hunk 14 above)
Comment (ntodosiychuk) [inline view_acl/tests/test_utils.py]: What about define variable with excluded modules?

Comment (ntodosiychuk) [inline view_acl/tests/test_utils.py]: Or should we handle this case by some verification if this module is system?

File: view_acl/tests/test_utils.py (hunk 8 above)
Comment (pzdeb) [inline view_acl/tests/test_utils.py]: our permissions are unique when composed from (app_name, view_name), where:
* app_name is the application the view is related to
* view_name is name of view (class or module) or value of `name` parameter if specified

It's completely ok when user use names for his views and even call a views group as a single `name` - they will share single permission codename in that case. That is a feature !?

File: view_acl/tests/test_utils.py (hunk 7 above)
Comment (pzdeb) [inline view_acl/tests/test_utils.py]: In my understanding `generate_permissions()` will return unique list of automatically generated permission `codenames`. All that name will be ready for storing to database

**Comment by pzdeb** [comment]
Let's format our permissions in such way:
`<PERM_PREFIX>_<app_name>_<view_name>`
In order to keep uniqueness for `view_name` we can combine `<app_name>_<view_name>` just to full path of the view (package + module + view name -- all lowercase)

File: view_acl/utils.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+
+
+def is_allowed_view(perm):
+    """Check if permission is in acl list."""
+
+    # Check if permission is in excluded list
+    for view in ACL_EXCLUDED_VIEWS:
+        module, separator, view_name = view.partition('*')
+        if view and perm.startswith(module):
+            return
```
Comment (pzdeb) [inline view_acl/utils.py]: looks like boolean result is missed for the return

File: view_acl/utils.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    return permissions
+
+
+def generate_perm_name(func):
+    """Generate permission name from callback function."""
+    return '.'.join((func and func.__module__ or '',
+                     func and func.__name__ or ''))
+
+
+def is_allowed_view(perm):
```
Comment (pzdeb) [inline view_acl/utils.py]: By default must return False

**Pull request merged**